thrift.go

```diff
@@ -0,0 +1,169 @@
+	log.InitLogger(level)
+}
+
+func initSecrets(ctx context.Context, cfg ServerConfig, logger log.Wrapper) (*secrets.Store, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*30)
+	defer cancel()
+
+	secretsStore, err := secrets.NewStore(ctx, cfg.Secrets.Path, logger)
+	if err != nil {
+		return nil, err
+	}
+	return secretsStore, nil
+}
+
+func initTracing(cfg ServerConfig, logger log.Wrapper, metrics *metricsbp.Statsd) error {
+	if err := tracing.InitGlobalTracer(tracing.TracerConfig{
+		ServiceName:      cfg.Tracing.Namespace,
+		SampleRate:       cfg.Tracing.SampleRate,
+		Logger:           logger,
+		MaxRecordTimeout: cfg.Tracing.RecordTimeout,
+		QueueName:        "main",
+	}); err != nil {
+		return err
+	}
+
+	tracing.RegisterCreateServerSpanHooks(
+		metricsbp.CreateServerSpanHook{Metrics: metrics},
+		tracing.ErrorReporterCreateServerSpanHook{},
+	)
+	return nil
+}
+
+func initSentry(cfg ServerConfig) error {
+	if err := raven.SetDSN(cfg.Sentry.DSN); err != nil {
+		return err
+	}
+	if err := raven.SetSampleRate(float32(cfg.Sentry.SampleRate)); err != nil {
+		return err
+	}
+	raven.SetEnvironment(cfg.Sentry.Environment)
+	return nil
+}
+
+func cleanup(closers []io.Closer) {
+	for _, c := range closers {
+		if err := c.Close(); err != nil {
+			log.Errorw("msg", "err", err)
+		}
+	}
+}
+
+// NewThriftServer returns a server that initializes and includes the default
+// middleware and cross-cutting concerns needed in order to be a standard Baseplate service.
+//
+// At the moment, this includes secrets management, metrics, edge contexts
+// (edgecontext.InjectThriftEdgeContext), and spans/tracing (tracing.InjectThriftServerSpan).
+func NewThriftServer(ctx context.Context, cfg ServerConfig, processor thriftbp.BaseplateProcessor, additionalMiddlewares ...thriftbp.ProcessorMiddleware) (srv Server, err error) {
+	var closers []io.Closer
+	defer func() {
+		if err != nil {
+			cleanup(closers)
+		}
+	}()
+
+	initLogger(cfg)
+	logger := log.ZapWrapper(cfg.Log.Level)
+	bootstrapLogger := log.ZapWrapper(log.ErrorLevel)
+
+	metricsbp.M = metricsbp.NewStatsd(ctx, metricsbp.StatsdConfig{
+		CounterSampleRate:   &cfg.Metrics.CounterSampleRate,
+		HistogramSampleRate: &cfg.Metrics.HistogramSampleRate,
+		Prefix:              cfg.Metrics.Namespace,
+		Address:             cfg.Metrics.Endpoint,
+		LogLevel:            cfg.Log.Level,
+	})
+	closers = append(closers, metricsbp.M)
+
+	secretsStore, err := initSecrets(ctx, cfg, bootstrapLogger)
+	if err != nil {
+		return nil, err
+	}
+	closers = append(closers, secretsStore)
+
+	ecImpl := edgecontext.Init(edgecontext.Config{Store: secretsStore})
+	if err = initTracing(cfg, bootstrapLogger, metricsbp.M); err != nil {
+		return nil, err
+	}
+	closers = append(closers, opentracing.GlobalTracer().(*tracing.Tracer))
+	if err = initSentry(cfg); err != nil {
+		return nil, err
+	}
+	innerCfg := thriftbp.ServerConfig{
+		Addr:    cfg.Addr,
+		Timeout: cfg.Timeout,
+		Logger:  logger,
+	}
```

Comment:
Same deal, I think all of this should be handled by a `Baseplate` struct possibly or some top level function, this feels like it's trying to do too much in one place.  I don't think we want to take the size of `main` methods down to _zero_.

Reply:
Generally, we do hope to take the main methods towards zero for the common cases and for the hello worlds.  Having a main that does a little too much, but that you can decompose 10% of the time for your advanced case, is ideal for gracefully onboarding new people to the system.